File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/import/removeUnusedImports.ts
function removeComments(body: string): string {
    return body
        .replace(/\/\/.*/g, '')
        .replace(/\/\*.*?\*\//gs, '')
}

export function removeUnusedImports(imports: string[], body: string): string[] {
    const cleanedBody = removeComments(body)

    return imports.filter(importItem => {
        if (importItem.includes("unhandled import")) return true

        const matcher = importItem.includes(" as ")
            ? /.+ as (.+)$/
            : /.+\.(.+)$/

        const match = importItem.match(matcher)

        if (!match) throw new Error(`Incorrect import: ${importItem}`)

        const [, name] = match

        return new RegExp(`\\b${name}\\b`).test(cleanedBody)
    })
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/import/removeUnusedImports.kt
package io.github.sgrishchenko.karakum.structure.import

import js.array.ReadonlyArray

private fun removeComments(body: String): String {
    return body
        .replace("//.*".toRegex(), "")
        .replace("/\\*(.|\n)*?\\*/".toRegex(RegexOption.MULTILINE), "")
}

fun removeUnusedImports(imports: ReadonlyArray<String>, body: String): ReadonlyArray<String> {
    val cleanedBody = removeComments(body)

    return imports.filter { importItem ->
        if ("unhandled import" in importItem) return@filter true

        val matcher = if (" as " in importItem) {
            ".+ as (.+)$".toRegex()
        } else {
            ".+\\.(.+)$".toRegex()
        }

        val match = matcher.find(importItem) ?: error("Incorrect import: $importItem")

        val (_, name) = match.groupValues

        "\\b${name}\\b".toRegex().containsMatchIn(cleanedBody)
    }.toTypedArray()
}
